### PR TITLE
check salable quantity for swatches are crossed out issue #33799

### LINKED ISSUE #33799: Swatches are not crossed out if salable quantity is 0
### Preconditions (*)

1. Magento 2,4,3
2.

### Steps to reproduce (*)

1. Make a configurable product with f.e. 3 options
2. grafik
3. Place an order for this product with one of the options so salable quantity is 0 after the order is placed. In the sample of the screenshot it is the product with option 0.1l and qty 3
4. After this order, in backend salable qty is 0 for the 0.1l product
5. grafik
6. But in frontend, the swatch for 0.1l is still selectable and not crossed out as it should.
7. grafik

### Expected result (*)

1. The swatch option 0.1l should be crossed and not selectable anymore as it worked in Mangeot 2.3.x
2.

### Actual result (*)

1. The swatch option 0.1l is not crossed and still seleactable.
2.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Kannakiraj123**
### Description (*)
Swatches are not crossed out if salable quantity is 0 and Product's stock status as "Is in stock".

### Related Pull Requests
1. I added "Magento\InventorySalesAdminUi\Model\GetSalableQuantityDataBySku" dependency injection to check salable qty at line No :102  in configurable product's helper data file.Function name (getOptions).In this file its checked only out of stock only.
2. please find below code changes at line no:101
                      $salableqty = $this->getSalableQuantityDataBySku->execute($product->getSku());
                       if ($product->isSalable() && (int)$salableqty[0]['qty']) {
                        ...
                        }

### Fixed Issues (if relevant)
1. Fixes magento/magento2#33799

### Manual testing scenarios (*)
1. Make a configurable product
2. Place an order for configurable product with one of the options so salable quantity is 0 after the order is placed.
3. After this order, in backend salable qty is 0 for respective option's product
4. But in frontend, the swatch will crossed out.

### CODE REVIEW

**Comment by Kannakiraj123** [comment]
@magento give me test instance

**Comment by Kannakiraj123** [comment]
@magento give me 2.4.develop_confiugrable_bugfix instance

**Comment by Kannakiraj123** [comment]
@magento run Unit Tests

File: app/code/Magento/ConfigurableProduct/Helper/Data.php [hunk 1]
```diff
@@ -90,7 +98,8 @@ public function getOptions($currentProduct, $allowedProducts)
                 $productAttribute = $attribute->getProductAttribute();
                 $productAttributeId = $productAttribute->getId();
                 $attributeValue = $product->getData($productAttribute->getAttributeCode());
-                if ($product->isSalable()) {
+                $salableqty = $this->getSalableQuantityDataBySku->execute($product->getSku());
```
Comment (kandy) [inline app/code/Magento/ConfigurableProduct/Helper/Data.php]: This implementation is not fully correct:
 - what if stack management is disabled
 - what if back orders is enabled
 
 Also , ConfigurableProduct must not depend on InventorySalesAdminUI module

**Comment by chequille** [comment]
Just one comment from my site. Tested it on my site.
Cache must be cleaned before the swatch is crossed. Therefore not the solution as requested.
Just one hint for it: it did work on Magento 2.3.5
Regards,
Chequille

**Comment by chequille** [comment]
Good Morning Everybody,
my tests are showing more things which are wrong in 2.4.3 release.
In magento 2.3.5 my configurable product with three options which do have all a salable quantity of 0 looks like the screenshot below. Swatches are not shown anymore and of course add to cart button is not shown:
grafik

In 2.4.3 release, unfortunanetly with the patch from above, it is still showing the swatches (all crossed now) and the add to cart button as well.
grafik

Again, please have a look at version 2.3.5 where everything does work as it should. We need to have this in 2.4.3 as well, as it is mandatory for going online with a shop.

Best regards,
Chequille

**Comment by Kannakiraj123** [comment]
hi @chequille ,I checked in my local env.Sorry but its not working test instance.Have to check.

**Comment by Kannakiraj123** [comment]
@magento give me 2.3.5 instance

**Comment by Kannakiraj123** [comment]
@magento give me 2.3.5-p2  instance

**Comment by Kannakiraj123** [comment]
@magento give me 2.3.5 instance

**Comment by chequille** [comment]
Hi @Kannakiraj123
it is really strange why this stock issues are not working anymore in 2.4.x.
I had as well a look to the code in 2.3.5 but I am not that deep in Magento2 to find the reason why it is working there.

Nevertheless, your idea with the observer is working for simple products with the code you made.
I made some changes in it to get it working as well for configurables. Of course I do not know if it is good code as of the rules.
I made as well a change in the Helper Data.php.

Changes/additions in the observer:
```

            if ($product->getTypeId() == 'configurable' )
            {	
				// if product is configurable go over all child products to see if product is still salable
            	$Childs = $product->getTypeInstance()->getUsedProducts($product);
            	foreach ( $Childs as $child) {
            		$stockItemConfig = $this->stockItemRepository->get($child->getId());
            		$stockItemData = $this->getStockItemData->execute($child->getSku(), $stockItemConfig->getStockId());
            		$minQty = $stockItemConfig->getMinQty();
            		$productQtyInStock += $stockItemData['quantity']
            		+ $this->getReservationsQuantity->execute($child->getSku(), $stockItemConfig->getStockId())
            		- $minQty;
            	}
				// If all option products do not have available stock set parent product to false
            	if( $productQtyInStock <=0 )
            		$observer->getEvent()->getSalable()->setIsSalable(false);
            		
            }
```
And change in Data.php
```
				if ($this->canDisplayShowOutOfStockStatus()) {
					if ($product->isSalable()) {
						// removed ['salable'] from this line:  $options['salable'][$productAttributeId][$attributeValue][] = $productId;
						$options[$productAttributeId][$attributeValue][] = $productId;
					}
					//removed this line: $options[$productAttributeId][$attributeValue][] = $productId;
				} else {
					if ($product->isSalable()) {
						$options[$productAttributeId][$attributeValue][] = $productId;
					}
				}
```

**Comment by Kannakiraj123** [comment]
hi @chequille ,Thanks for coding updation.yes we have to check 2.3.5.Still am checking which difference make issue 2.4.x.

**Comment by Kannakiraj123** [comment]
hi @chequille ,The difference of 2.4.x and 2.3.5 is one file "Magento\InventoryIndexer\Model\IsProductSalable".I changed code of this file.please check once this file.

**Comment by Kannakiraj123** [comment]
@magento give me test instance

**Comment by Kannakiraj123** [comment]
please confirm in git clone wont give "Magento_InventoryIndexer" module.is it correct to add file in my branch?

**Comment by chequille** [comment]
@Kannakiraj123 
Ok, I removed the observer and made the changes in InventoryIndexer modul and I kept my changes in the Data.php of the configurable product, because your code is not crossing the swatches.
Results of my tests:
1. ADD to CART button is not shown anymore if the product does have Salable Quantity of 0 in catalog list/grid view and in the product view.
2. Swatches are crossed if Salable Quantity is 0 but there is the need of cleaning the cache before

So far my test. It seems that this change in the InventoryIndexer solves as well other issues I made in the past which depend on the IsSalable() function which always returns true if stock is >0 and not if salable qty is > 0.

Have a nice rest of the weekend.
BR
Chequille

**Comment by Kannakiraj123** [comment]
@magento give me test instance

**Comment by Kannakiraj123** [comment]
@magento run all tests

**Comment by slavvka** [comment]
@magento give me test instance

**Comment by chequille** [comment]
What is going on here?
Actual status of this is, that it is working if product view page is not cached.
So, this has to be solved.
Any ideas?

**Comment by gjportegies** [comment]
> Hi @Kannakiraj123 it is really strange why this stock issues are not working anymore in 2.4.x. I had as well a look to the code in 2.3.5 but I am not that deep in Magento2 to find the reason why it is working there.
> 
> Nevertheless, your idea with the observer is working for simple products with the code you made. I made some changes in it to get it working as well for configurables. Of course I do not know if it is good code as of the rules. I made as well a change in the Helper Data.php.
> 
> Changes/additions in the observer:
> 
> ```
> 
>             if ($product->getTypeId() == 'configurable' )
>             {	
> 				// if product is configurable go over all child products to see if product is still salable
>             	$Childs = $product->getTypeInstance()->getUsedProducts($product);
>             	foreach ( $Childs as $child) {
>             		$stockItemConfig = $this->stockItemRepository->get($child->getId());
>             		$stockItemData = $this->getStockItemData->execute($child->getSku(), $stockItemConfig->getStockId());
>             		$minQty = $stockItemConfig->getMinQty();
>             		$productQtyInStock += $stockItemData['quantity']
>             		+ $this->getReservationsQuantity->execute($child->getSku(), $stockItemConfig->getStockId())
>             		- $minQty;
>             	}
> 				// If all option products do not have available stock set parent product to false
>             	if( $productQtyInStock <=0 )
>             		$observer->getEvent()->getSalable()->setIsSalable(false);
>             		
>             }
> ```
> 
> And change in Data.php
> 
> ```
> 				if ($this->canDisplayShowOutOfStockStatus()) {
> 					if ($product->isSalable()) {
> 						// removed ['salable'] from this line:  $options['salable'][$productAttributeId][$attributeValue][] = $productId;
> 						$options[$productAttributeId][$attributeValue][] = $productId;
> 					}
> 					//removed this line: $options[$productAttributeId][$attributeValue][] = $productId;
> 				} else {
> 					if ($product->isSalable()) {
> 						$options[$productAttributeId][$attributeValue][] = $productId;
> 					}
> 				}
> ```
> 
> This would be now a workaround for the problem. Of course the better way would be to find the reason why it was OK in 2.3.5 and not OK in 2.4.x And of course bundle products must be included as well. ANd as well cache has to be looked at.
> 
> Regards, Chequille

Hi @chequille in which exact file did you apply this changes? Possible to give a full example?

**Comment by chequille** [comment]
Hi @gjportegies,

currently, I made the following to get it working acceptable.

In File magento-configurable-product/Helper/Data.php and the method getOptions

`	/**
	 * Get Options for Configurable Product Options
	 *
	 * @param Product $currentProduct
	 * @param array $allowedProducts
	 * @return array
	 */
	public function getOptions($currentProduct, $allowedProducts)
	{
		$options = [];
		$allowAttributes = $this->getAllowAttributes($currentProduct);
		
		foreach ($allowedProducts as $product) {
			$productId = $product->getId();
			foreach ($allowAttributes as $attribute) {
				$productAttribute = $attribute->getProductAttribute();
				$productAttributeId = $productAttribute->getId();
				$attributeValue = $product->getData($productAttribute->getAttributeCode());
				if ($this->canDisplayShowOutOfStockStatus()) {
					if ($product->isSalable()) {
						$options[$productAttributeId][$attributeValue][] = $productId;
						// Original code line:  $options['salable'][$productAttributeId][$attributeValue][] = $productId;
					}
					// Original code line: $options[$productAttributeId][$attributeValue][] = $productId;
				} else {
					if ($product->isSalable()) {
						$options[$productAttributeId][$attributeValue][] = $productId;
					}
				}
				$options['index'][$productId][$productAttributeId] = $attributeValue;
			}
		}
		$options['canDisplayShowOutOfStockStatus'] = $this->canDisplayShowOutOfStockStatus();
		return $options;
	}
`
I removed one line and modified the other one as above.

And I use the complete file magento/inventory-indexer/Model/IsProductSalable.php

`<?php

/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
declare(strict_types=1);

namespace Magento\InventoryIndexer\Model;

use Magento\Framework\Exception\LocalizedException;
use Magento\InventorySalesApi\Api\IsProductSalableInterface;
use Magento\InventorySales\Model\IsProductSalableCondition\IsProductSalableConditionChain;
use Psr\Log\LoggerInterface;

/**
 * Lightweight implementation for Storefront application.
 */
class IsProductSalable implements IsProductSalableInterface
{
	/**
	 * @var IsProductSalableConditionChain
	 */
	private $isProductSalableConditionChain;
	/**
	 * @var LoggerInterface
	 */
	private $logger;
	
	/**
	 * @param IsProductSalableConditionChain $getStockItemData
	 * @param LoggerInterface $logger
	 */
	public function __construct(
			IsProductSalableConditionChain $isProductSalableConditionChain,
			LoggerInterface                $logger
			)
	{
		$this->isProductSalableConditionChain = $isProductSalableConditionChain;
		$this->logger = $logger;
	}
	
	/**
	 * @inheritdoc
	 */
	public function execute(string $sku, int $stockId): bool
	{
		try {
			$isSalable = $this->isProductSalableConditionChain->execute($sku, $stockId);
		} catch (LocalizedException $exception) {
			$this->logger->warning(
					sprintf(
							'Unable to fetch stock #%s data for SKU %s. Reason: %s',
							$stockId,
							$sku,
							$exception->getMessage()
							)
					);
			$isSalable = false;
		}
		
		return $isSalable;
	}
}`

The only thing I must do additionlly is to make my product page or catalog page non cachable, because otherwise it is only working if you clear the cache each time there is a change.
I made this setting one block in catalog_product.view.xml non cachable.

This is how it works for me currently, even if it is just a workaround as none of the above changes in the core files are commited by the developers of Magento. They just do not care about this massive problem with stock and cache, unfortunately.

I hope it helps you.
Best regards,
Jürgen

**Comment by Kannakiraj123** [comment]
@magento give me 2.4.develop instance

**Comment by Kannakiraj123** [comment]
@magento run all tests

**Pull request closed without merging**